aten/src/ATen/ThreadLocalState.cpp

```diff
@@ -6,35 +6,96 @@
 
 namespace at {
 
+namespace {
+thread_local bool is_record_function_enabled_ = true;
+typedef std::array<std::function<SettingValue()>, (size_t)ThreadLocalSetting::NUM_SETTINGS> getters_arr;
+getters_arr& getters() {
+  static getters_arr getters;
+  return getters;
+}
+
+typedef std::array<std::function<void(SettingValue)>, (size_t)ThreadLocalSetting::NUM_SETTINGS> setters_arr;
+setters_arr& setters() {
+  static setters_arr setters;
+  return setters;
+}
+
+bool _unused = []() {
+  ThreadLocalState::registerThreadLocalSetting(
+    ThreadLocalSetting::GRAD_MODE,
+#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```

Comment:
nit: can't you move it to some of the autograd files so that you can avoid an ugly ifdef?

Reply:
otoh, we may need have header on mobile platform, so we would still need an ifdef

Reply:
yeah, you kind of have the same ifdef in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/grad_mode.cpp already. We could move this registration there, but I guess it's about the same